Enhanced ping monitor with advanced options (count, timeout, numeric)

filippolauria opened the pull request:
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description
This PR enhances ping monitoring capabilities by introducing advanced configuration options that give users more control over ping behavior:

- Max Packets (count): Specify how many packets to send before stopping
- Maximum Duration (deadline): Set total time limit for the ping operation
- Response Timeout (timeout): Control how long to wait for each packet response
- Numeric Output (numeric): Choose between IP addresses or hostnames in ping output

These options match the standard ping command features, making the monitoring system more flexible and suitable for different network scenarios. Each option has been implemented with proper validation boundaries and default values.

This PR is kinda related to issue #5552 
Resolves #5785

## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

![image](https://github.com/user-attachments/assets/4bedc7e7-e90f-4d07-a6db-dae6ecba7daf)

filippolauria commented:
Hi @louislam and other mainteiners!

Just checking in on this PR.

All checks are passing and I'm happy to address any feedback or concerns you might have.

CommanderStorm commented on db/knex_init_db.js:
```diff
@@ -548,6 +548,20 @@ ALTER TABLE monitor
+    ALTER TABLE monitor ADD ping_deadline INTEGER default 10 not null;
+    ALTER TABLE monitor ADD ping_timeout INTEGER default 2 not null;
+    */
+    await knex.schema.table("monitor", function (table) {
```
please remove this part and add it to a [new migration](https://github.com/louislam/uptime-kuma/tree/003ffc2cc2a3b572c197658b08b34a6618f804fa/db/knex_migrations), and only there (the `.sql` files are a crutch which we did not have the time/energy to clean up yet)

I think this file will only be run for new users, but not existing users.

CommanderStorm commented on db/knex_init_db.js:
```diff
@@ -548,6 +548,20 @@ ALTER TABLE monitor
+        table.integer("ping_count").defaultTo(1).notNullable();
+        table.boolean("ping_numeric").defaultTo(true).notNullable();
+        table.integer("ping_deadline").defaultTo(10).notNullable();
+        table.integer("ping_timeout").defaultTo(2).notNullable();
```
please re-use the existing timeout field instead of adding a new one

CommanderStorm commented on server/util-server.js:
```diff
@@ -118,20 +122,33 @@ exports.tcping = function (hostname, port) {
+ * @param {string} sourceAddr Source address for sending/receiving echo requests
+ * @param {boolean} numeric If true, IP addresses will be output instead of symbolic hostnames
+ * @param {number} size Size (in bytes) of echo request to send
+ * @param {number} deadline Maximum time in seconds before ping stops, regardless of packets sent
```
I think we can remove the deadline option and fold it into the timeout option.
Having to explain to users the difference between deadline and timeout would be difficult (I also don't just looking at the code)

filippolauria replied:
In the advanced section of a monitoring tool, I would expect to find these kinds of configurations, which in this case come directly from the [ping manual](https://linux.die.net/man/8/ping).

Simply put, `timeout` defines how long to wait for a reply after sending a request, while `deadline` sets the total execution time of the ping process. For example, if you configure 10 pings with a 2-second `timeout`, in the worst case (no replies received), the process would take 20 seconds to complete. However, if you also set a `deadline` of 10 seconds, the ping process will stop after 10 seconds, regardless of the number of pings sent or replies received.

To make this distinction clearer in the interface:

- `deadline` is labeled as _Max duration_, with the description: _Total time in seconds before ping stops, regardless of packets sent_  
- `timeout` is labeled as _Response timeout_, with the description: _Maximum time in seconds to wait for each response_

I hope this explanation clarifies the difference. More details can be found in the ping manual (linked above).

PS: The order of the options in the advanced section wasn’t chosen randomly; it follows the order in the ping manual.

CommanderStorm replied:
Please change this to our definition of a timeout, like for the other monitors:
Timeout is the max. Time after which the monitor fails

The other could likely be named "per-ping timeout".
Having two timeouts 

I don't think that this is a case where this monitor needs to be special

CommanderStorm commented on src/pages/EditMonitor.vue:
```diff
@@ -660,10 +660,48 @@
+                                    {{ $t("Numeric Output") }}
+                                </label>
+                                <div class="form-text">
+                                    {{ $t("If checked, IP addresses will be output instead of symbolic hostnames") }}
```
please make sure that all translations are added to `en.json`.
Otherwise they are not avaliable to our translators

CommanderStorm commented on src/pages/EditMonitor.vue:
```diff
@@ -1133,6 +1188,14 @@ export default {
         return {
             minInterval: MIN_INTERVAL_SECOND,
             maxInterval: MAX_INTERVAL_SECOND,
+            ping_count_min: PING_COUNT_MIN,
+            ping_count_max: PING_COUNT_MAX,
+            packetSize_min: PING_PACKET_SIZE_MIN,
+            packetSize_max: PING_PACKET_SIZE_MAX,
+            ping_deadline_min: PING_DEADLINE_MIN,
+            ping_deadline_max: PING_DEADLINE_MAX,
+            ping_timeout_min: PING_TIMEOUT_MIN,
+            ping_timeout_max: PING_TIMEOUT_MAX,
```
I would like for this not being a `data`.

Most other cases where this is happeinng in this project (except for interval) choose to inline these values and duplicate them in this circumstance between front and backend.
Currently the deduplication makes is a bit hard to read what all checks are.

=> lets inline these values as well here. This has the benefit of also making this simper to review

CommanderStorm commented on server/util-server.js:
```diff
@@ -140,18 +157,34 @@ exports.ping = async (hostname, size = 56) => {
+    ipv6 = false,
+    count = PING_COUNT_DEFAULT,
+    sourceAddr = "",
+    numeric = true,
```
Please re-test if `numeric` works as expected. 
It is not used below

filippolauria replied:
Addressed

CommanderStorm reviewed (changes requested):
I had a quick glance over the code and commented where things need a bit more work.
Seems overall reasonable to me.
I have not had time to properly test that all code works as advertised ^^.

I don't like how messy the code for `EditMonitor` is becoming and that whole verifying parameters 2x and in different places is not great..
This is not your fault and should not be fixed in this PR.

You also stumbled into the big `old_migrations` pit.. sorry for that

dredvard commented:
Does this pr have the ability to satisfuy this feature request? 
https://github.com/louislam/uptime-kuma/issues/4234

filippolauria commented:
@dredvard This PR maps some _advanced_ options (`-c`, `-n`, `-w`, `-W`, as referenced in the [ping manual](https://linux.die.net/man/8/ping)) into the graphical interface. Of course, these options have always been available in the ping tool but were never included in the UI until now (hopefully).

CommanderStorm reviewed (changes requested):
I am still not happy with the third timeout field. This is not something that requires being "special" => let's not be special

filippolauria commented:
@CommanderStorm ,
I have removed the `deadline` field and changed the existing `timeout` field to represent a _global timeout_ (ping's -w option). Also, I have introduced the `ping_per_request_timeout` field for _individual ICMP echo request timeout_ (ping's -W option).

CommanderStorm commented on db/knex_migrations/2025-03-04-0000-ping-advanced-options.js:
```diff
@@ -0,0 +1,24 @@
+        .alterTable("monitor", function (table) {
+            table.integer("ping_count").defaultTo(1).notNullable();
+            table.boolean("ping_numeric").defaultTo(true).notNullable();
+            table.integer("ping_deadline").defaultTo(10).notNullable();
```
@filippolauria please re-read my comment. I think there might have been a misunderstanding.

Existing monotors now get `ping_per_sub_ping_timeout=2`, as not specified beforehand.
So this might be a a breaking change behaviour wise => would need to wait for V3.0, as we have already released the first beta of v2.0..

This would be resolved if you were to migrate them to `ping_per_sub_ping_timeout=timeout` for existing monitors

filippolauria replied:
@CommanderStorm thank you for this clarification. I will use `timeout` as per-ping timeout. Then I'll remove the extra field (the `deadline`) and use a percentage of the retry interval as the actual deadline (e.g. 90%). Could this be ok?

CommanderStorm replied:
That sounds unnecessarily complicated.
Why not just migrate existing values to not cause breakage?

(In knex that can be done via `.then(..)`)

filippolauria replied:
Sorry, I've had the chance to carefully review what you're saying.

**TL;DR Actually, at the moment the `timeout` field is not used in ping monitor**

---
The ping monitor relies on [node-ping](https://github.com/louislam/node-ping). Throughout uptime-kuma just a few parameters can be actually chosen by the user, when using this lib (namely: hostname and packetSize)
https://github.com/louislam/uptime-kuma/blob/a7d1b99719c70263e0509cec708bdd759af51e2a/server/model/monitor.js#L620
So at the moment, there are several default hardcoded values that statically apply to ping monitor (eg. look at this linux builder: https://github.com/louislam/node-ping/blob/09ad252663c4104f2841c624081174ae568fed2a/lib/builder/linux.js#L33-L40) which I am trying to parameterize
https://github.com/louislam/uptime-kuma/blob/fb499c8a9cddfc18bd9c41543c81e392e5e66c79/server/model/monitor.js#L637


So initializing global timeout, per-ping timeout, numeric and count to default values seems the more appropriate choice.

---

Was this your initial concern?

CommanderStorm reviewed (approved):
LGTM from my side then.
